edb/lang/graphql/translator.py

```diff
@@ -459,12 +464,68 @@ def _validate_fragment_type(self, frag, spread):
     def _visit_arguments(self, arguments):
         where = offset = limit = None
         orderby = []
+        first = last = before = after = None
 
         for arg in arguments:
-            if arg.name == 'filter':
-                where = self.visit(arg.value)
-            elif arg.name == 'order':
-                orderby = self.visit_order(arg.value)
+            try:
+                if arg.name == 'filter':
+                    where = self.visit(arg.value)
+                elif arg.name == 'order':
+                    orderby = self.visit_order(arg.value)
+                elif arg.name == 'first':
+                    first = arg.value.value
+                    if first < 0:
+                        raise ValueError(f"{arg.name!r} cannot be negative")
+                elif arg.name == 'last':
+                    last = arg.value.value
+                    if last < 0:
+                        raise ValueError(f"{arg.name!r} cannot be negative")
+                elif arg.name == 'before':
+                    before = int(arg.value.value)
+                    if before < 0:
+                        raise ValueError(f"{arg.name!r} cannot be negative")
+                elif arg.name == 'after':
+                    after = int(arg.value.value)
+                    if after < 0:
+                        raise ValueError(f"{arg.name!r} cannot be negative")
+                    # The +1 is to make 'after' into an appropriate index.
+                    #
+                    # 0--a--1--b--2--c--3-- ... we call element at
+                    # index 0 (or "element 0" for short), the element
+                    # immediately after the mark 0. So after "element
+                    # 0" really means after "index 1".
+                    after += 1
+            except Exception:
+                raise GraphQLValidationError(
+                    f"invalid value for {arg.name!r}: {arg.value.value!r}",
+                    context=arg.context)
+
+        # convert before, after, first and last into offset and limit
+        if after is not None:
+            offset = after
+        if before is not None:
+            limit = before - (after or 0)
+        if first is not None:
+            if limit is None:
+                limit = first
+            else:
+                limit = min(first, limit)
+        if last is not None:
+            if limit is not None:
+                if last < limit:
+                    offset = (offset or 0) + limit - last
+                    limit = last
+            else:
+                # FIXME: there wasn't any limit, so we can define last
+                # in terms of offset alone without negative OFFSET
+                # implementation
+                raise NotImplementedError
```

Comment:
I suggest to raise the newly added `GraphQLTranslationError('negative offsets are not supported')`

Reply:
This isn't accurate. This exception would be raised on otherwise legal GraphQL (e.g. argument `last: 3`). Technically the argument is fine according to the spec and semantically valid. It is not implemented because ideally we'd rather implement it using EdgeQL `OFFSET -3`. That is an implementation details, though. Negative `OFFSET` is an EdgeQL feature and it currently is not implemented. The end user doesn't necessarily know anything about negative offsets.

Reply:
Sure, then raise 

```python
raise GraphQLTranslationError(
    f'limit={limit} and offset={offset}  translate to a negative OFFSET in EdgeQL '
    f'which is currently unsupported')
```

Reply:
I mean this isn't something that we'll have for a long time, but if someone encounters this exception it's good that they will know roughly what's going on (and that the error isn't their fault).